lib/tree.rb

```diff
@@ -16,42 +16,100 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(log n) 
+  # Space Complexity: O(1)
   def add(key, value)
-    raise NotImplementedError
+    new_node = TreeNode.new(key, value)
+
+    if @root ==  nil
+      @root = new_node
+    else
+      add_node(@root, new_node)
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def add_node(root, new_node)
+    if new_node.key < root.key
+      return root.left = new_node unless root.left
+      add_node(root.left, new_node)
+    elsif new_node.key > root.key
+      return root.right = new_node unless root.right
+      add_node(root.right, new_node)
+    end
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
   def find(key)
-    raise NotImplementedError
+    return nil if @root == nil
+
+    current = @root
+
+    until current == nil
+      if key == current.key
+        return current.value
+      elsif key < current.key
+        current = current.left
+      else
+        current = current.right
+      end
+    end
+
+    return nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def inorder
-    raise NotImplementedError
+    return [] if @root == nil
+
+    current = @root 
+    tree = []
+      
+    return inorder_helper(current, tree)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def preorder
-    raise NotImplementedError
+    return [] if @root == nil
+
+    current = @root 
+    tree = []
+      
+    return preorder_helper(current, tree)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def postorder
-    raise NotImplementedError
+    return [] if @root == nil
+
+    current = @root 
+    array = []
+      
+    return postorder_helper(current, array)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
```

Comment:
Since you are doing recursion I would say this is O(log n) where n is the number of elements ,assuming it's a balanced tree.   Remember doing recursion requires a stack!